### PR TITLE
postgres module uuid-oosp documentation

### PR DESCRIPTION
**Pull request by OOlsen**
* add a new prerequisites section with db version and how to provide
that module
* refer to the prerequisites guide in the migration guide

I chose 9.4 as the min postgres version, as there was the jsonb feature introduced.  It' the newest postgres feature used by the factcast as far as i know.

### CODE REVIEW

**Comment by OOlsen** [comment]
The suggestion is to switch from the superuser to a user with less prviliges after the initial start. I'm a litte bit concerned that the "final" user miss some privilges as he is not the owner of db objects.
But According to the Postgresql [documentation](https://www.postgresql.org/docs/11/ddl-schemas.html) it should work:
"Note that by default, everyone has CREATE and USAGE privileges on the schema public. "

Maybe we should test if the current usage and further liquibase updates will success without the db object owner?

**Comment by uweschaefer** [comment]
shouldn't we do something like this?
https://stackoverflow.com/questions/26774525/liquibase-spring-how-to-use-separate-user-for-schema-changes

having a different user for DDL feels about right.

**Comment by uweschaefer** [comment]
will replace UUID returned by a dedicated Object and adjust documentation accordingly

**Comment by OOlsen** [comment]
System.setProperty("spring.datasource.username", "olsen");
     System.setProperty("spring.datasource.password", "olsen");
     System.setProperty("spring.liquibase.user", "postgres");
     System.setProperty("spring.liquibase.password", "postgres");

works with the minimum of privileges :

CREATE USER "olsen" WITH PASSWORD 'olsen';
GRANT SELECT, INSERT, DELETE, UPDATE ON TABLE fact, tokenstore, catchup TO olsen;
GRANT EXECUTE ON  FUNCTION notifyFactInsert TO olsen;
GRANT USAGE,SELECT ON SEQUENCE catchup_seq, fact_ser_seq TO olsen;

I will add a hint, that the user should change the password of the default application user after the initial start.

**Comment by OOlsen** [comment]
Now the liquibase user do not only need additional permissions to install the uuid module, it also has to take care of the application user creation and privileges assignment. We should document the necessary privileges for the liquibase user to provide the application user:
The alternative would be, we demand the usage of a superuser and remove the choice to install the uuid module manually?

**Comment by OOlsen** [comment]
I'm fine the the current state. @uweschaefer  can you pls take a look.
I added a new application user to liquibase and updated the docu.

**Comment by OOlsen** [comment]
That's strange. Tested it before and again right now.

`2019-05-01 20:20:40 [main] INFO  liquibase.executor.jvm.JdbcExecutor -
        --liquibase createApplicationUser sql
--changeset akp:issue372

CREATE USER "factcast-application-user" WITH PASSWORD 'factcast-application-user';
GRANT SELECT, INSERT, DELETE, UPDATE ON TABLE fact, tokenstore, catchup TO "factcast-application-user";
GRANT EXECUTE ON  FUNCTION notifyFactInsert TO "factcast-application-user";
GRANT USAGE,SELECT ON SEQUENCE catchup_seq, fact_ser_seq TO "factcast-application-user"
2019-05-01 20:20:40 [main] INFO  liquibase.changelog.ChangeSet -
        Custom SQL executed
2019-05-01 20:20:40 [main] INFO  liquibase.changelog.ChangeSet -
        ChangeSet classpath:/db/changelog/factcast/issue372/createApplicationUser.sql::raw::includeAll ran successfully in 14ms`

is part of may output and the server runs fine.

The only difference in our attempts is that i use the current postgres docker image from the docker registry instead of test-containers. as far as i see

**Comment by uweschaefer** [comment]
i guess there has been a change in the grant syntax, so that older PGs cannot parse the current sql string.
could you check the syntax, and tweak it, so that it runs with older PGs, too? (maybe just use the example server using the testcontainer as currently configured).

are the " corrrect? or should it be rather ' ? (just guessing)

**Comment by OOlsen** [comment]
Seems like the brackets for the function are mandatory for the testcontainer version to grant permissions. I made a sucessfull run with the currently configured testcontainer version from the example project and also with the version from docker registry.

btw. i got rid of the "connot connect to ruyk" testcontainer error by replacing the fedora-docker with the current docker-ce. It was a pretty old version.

**Comment by uweschaefer** [comment]
will have another look on the WE

File: factcast-site/documentation/content/migration/_index.md [hunk 1]
```diff
@@ -23,6 +23,14 @@ If you used the uniqe_identifier feature before, it was removed. It was only a r
 
 There is a [section on optimitic locking](/usage/java/optimistic_locking/) as a new api feature.
 
+#### Postgres module uuid-ossp and ddl user
+
+The Postgres module *uuid-ossp* is necessary for the new optimistic locking api feature. In order to provide a user with enough privileges to install modules and also running the application without unnecessary db permission, we split the database responsibility into two different users. 
+If you want to upgrade your server version, please take a look into the [{{%icon circle-arrow-right%}}Postgres]({{%relref "/setup/server/prerequisites.md#postgres"%}}) section in the prerequisites guide to ensure that your migration will succeed. 
+One Additional note: If you don't want to provide a superuser as _ddl user_, make sure that you provided user has the permissions to assign privileges for the already existing database objects to the new created _application user_.
```
Comment (mweirauch) [inline factcast-site/documentation/content/migration/_index.md]: I don't quite get the sentence... was this supposed to be something like "If you don't want to provide a superuser as _ddl user_, make sure that you**r** provided _application user_ has the permissions to assign privileges for the already existing database objects." ?

**Comment by mweirauch** [comment]
Overall, what is the reasoning of providing/creating a default "factcast-application-user"? From my understanding, it will be created always.

**Scenario:**
Given the case I am managing my database ddl-user (let's call him "my-db-admin") and dml-user ("my-factcast-user") externally and pass the required information to the application via externalized configuration on the first start.
After the application has been started the first time I will now have a default "factcast-application-user"  (next to my externally managed database users) whose password needs to be altered afterwards (because it's still the default) or the user has to be removed entirely manually because one doesn't need it.
Between the initial creation of the "factcast-application-user" and the removal of this user or his password change, there is an attack vector for those knowing about this "created-by-default" user credentials.

Am I missing something? Is this kind of a common pattern in other software? (Haven't seen so, that's why I am asking.)

**Comment by OOlsen** [comment]
If everybody provides the dml-user manually, we can remove the self created "factcast-application-user". Will that be the case?
The intention of this user is to remove administration complexity. You don't have to take care, that your manually provided dml-user has all the necessary permissions to fulfil its job. Additionally, it may have unnecessary permission. To provide the user by the application itself, exactly all needed permissions are provided.
But the implicit created application-user creates a vulnerable time window with its default password, that's true.
An alternative to the implicit created dml-user would be, that we document the need of an provided application user with all usage permission for current and future existing  database objects in the public schema.

File: factcast-site/documentation/content/migration/_index.md (hunk 1 above)
Comment (OOlsen) [reply factcast-site/documentation/content/migration/_index.md]: yes, thats what i meant

**Comment by mweirauch** [comment]
Perhaps one could solve this with [Changelog Parameters](http://www.liquibase.org/documentation/changelog_parameters.html) so that the dml-user can be specified externally? (Perhaps it's possible to reuse the `spring.datasource.username` and `spring.datasource.password` assuming the `ApplicationContext` is already initialized at this stage.)

**Comment by OOlsen** [comment]
Using the Changelog Parameter works very well. We can provide user and pw externally and can also make the creation conditional. The changeset will be skipped, if the user already exists.

But currently we need 2 additional parameters for that.
I'm still considering, if we should forward the spring.datasource params as changelog parameters to liquibase. There will be users that don't use the spring.datasource properties to create their datasources, so we have to document the new liquibase changelog parameters anyways. The liquibase properties are optional and we have to make clear who wins, if both the liquibase and the spring.datasource properties are set.
On the other hand, people who allready  using the spring.datasource properties doesn't have to take care of the additional liquibase properties, which will will only be needed one time, when the changeset will be applied the first time.

Without forwarding the spring.datasource credentials, the user configuration looks like the following example:

`
spring.datasource.username=dml-user
spring.datasource.password=dml-user-pw
spring.liquibase.parameters.dml-user=dml-user
spring.liquibase.parameters.dml-user-pw=dml-user-pw
spring.liquibase.user=ddl-user
spring.liquibase.password=ddl-user-pw
`

**Comment by uweschaefer** [comment]
@OOlsen  please have a look at 04d9745. this might be a viable alternative that prevents us from fiddeling with two users.

**Comment by uweschaefer** [comment]
crap. https://stackoverflow.com/questions/46981873/permission-denied-to-set-session-authorization-on-amazon-postgres-rds

**Comment by uweschaefer** [comment]
could anyone please try with RDS? cannot do it from home.

**Comment by uweschaefer** [comment]
In case SET SESSION AUTHORIZATION does not work on RDS, i suggest removing the user business and fall back to documentation only for a restricted user.
Liquibase is not run by an AutoConfiguration, but an ApplicationListener, which makes it hard to get in between.

**Comment by mweirauch** [comment]
The usual nitpick from my side: In case this is possible on RDS and the user creation is pursued, I'd have favoured keeping the "hypthen-case" username. (That "camelCaseUser" looks utterly unusual.)

**Comment by OOlsen** [comment]
made a test with rds.
The server starts without problems, but the server logs errors after client requests.

```
2019-05-08 12:23:43 [grpc-default-executor-2] ERROR o.a.t.jdbc.pool.PooledConnection -
        SQL Validation error
org.postgresql.util.PSQLException: ERROR: permission denied to set session authorization
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:308)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:307)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:293)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:270)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:266)
	at org.apache.tomcat.jdbc.pool.PooledConnection.validate(PooledConnection.java:543)
	at org.apache.tomcat.jdbc.pool.PooledConnection.validate(PooledConnection.java:454)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.createConnection(ConnectionPool.java:740)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.borrowConnection(ConnectionPool.java:671)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.getConnection(ConnectionPool.java:198)
	at org.apache.tomcat.jdbc.pool.DataSourceProxy.getConnection(DataSourceProxy.java:132)
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:262)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:378)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:474)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:289)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.factcast.store.pgsql.internal.PgFactStore$$EnhancerBySpringCGLIB$$f0bfe8e1.publish(<generated>)
	at org.factcast.server.grpc.FactStoreGrpcService.publish(FactStoreGrpcService.java:99)
	at org.factcast.server.grpc.FactStoreGrpcService$$FastClassBySpringCGLIB$$10df9000.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:749)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:69)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.factcast.server.grpc.FactStoreGrpcService$$EnhancerBySpringCGLIB$$9d3058f4.publish(<generated>)
	at org.factcast.grpc.api.gen.RemoteFactStoreGrpc$MethodHandlers.invoke(RemoteFactStoreGrpc.java:861)
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at net.devh.boot.grpc.server.security.interceptors.AuthenticatingServerInterceptor$AuthenticatingServerCallListener.onHalfClose(AuthenticatingServerInterceptor.java:151)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at net.devh.boot.grpc.server.security.interceptors.ExceptionTranslatingServerInterceptor$ExceptionTranslatorServerCallListener.onHalfClose(ExceptionTranslatingServerInterceptor.java:111)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:283)
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:711)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
2019-05-08 12:23:43 [grpc-default-executor-2] ERROR o.f.server.grpc.FactStoreGrpcService -
        Problem while publishing: 
org.springframework.transaction.CannotCreateTransactionException: Could not open JDBC Connection for transaction; nested exception is java.sql.SQLException: Validation Query Failed, enable logValidationErrors for more details.
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:305)
	at org.springframework.transaction.support.AbstractPlatformTransactionManager.getTransaction(AbstractPlatformTransactionManager.java:378)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.createTransactionIfNecessary(TransactionAspectSupport.java:474)
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:289)
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:98)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.factcast.store.pgsql.internal.PgFactStore$$EnhancerBySpringCGLIB$$f0bfe8e1.publish(<generated>)
	at org.factcast.server.grpc.FactStoreGrpcService.publish(FactStoreGrpcService.java:99)
	at org.factcast.server.grpc.FactStoreGrpcService$$FastClassBySpringCGLIB$$10df9000.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:749)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.security.access.intercept.aopalliance.MethodSecurityInterceptor.invoke(MethodSecurityInterceptor.java:69)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at org.factcast.server.grpc.FactStoreGrpcService$$EnhancerBySpringCGLIB$$9d3058f4.publish(<generated>)
	at org.factcast.grpc.api.gen.RemoteFactStoreGrpc$MethodHandlers.invoke(RemoteFactStoreGrpc.java:861)
	at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at net.devh.boot.grpc.server.security.interceptors.AuthenticatingServerInterceptor$AuthenticatingServerCallListener.onHalfClose(AuthenticatingServerInterceptor.java:151)
	at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:35)
	at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
	at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
	at net.devh.boot.grpc.server.security.interceptors.ExceptionTranslatingServerInterceptor$ExceptionTranslatorServerCallListener.onHalfClose(ExceptionTranslatingServerInterceptor.java:111)
	at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:283)
	at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:711)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.sql.SQLException: Validation Query Failed, enable logValidationErrors for more details.
	at org.apache.tomcat.jdbc.pool.ConnectionPool.createConnection(ConnectionPool.java:754)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.borrowConnection(ConnectionPool.java:671)
	at org.apache.tomcat.jdbc.pool.ConnectionPool.getConnection(ConnectionPool.java:198)
	at org.apache.tomcat.jdbc.pool.DataSourceProxy.getConnection(DataSourceProxy.java:132)
	at org.springframework.jdbc.datasource.DataSourceTransactionManager.doBegin(DataSourceTransactionManager.java:262)
	... 33 common frames omitted
```

**Comment by OOlsen** [comment]
I gave the "dml datasource credentials to liquibase forwarding" a try.
[5cef964cc46361340d095ea8690fd335bc4de443](https://github.com/Mercateo/factcast/commit/5cef964cc46361340d095ea8690fd335bc4de443).
It runs successful with rds autoconf and conventional spring.datasource properties.

**Comment by mweirauch** [comment]
I only tested the database credential handling. Works for me with a given "dml-user". (Given he has SUPERUSER rights to install the extension.)

Thanks!

File: factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/LiquibaseChangelogParamsForwarder.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    public Object postProcessBeforeInitialization(@NonNull Object bean, @NonNull String beanName)
+            throws BeansException {
+
+        if (bean instanceof SpringLiquibase) {
+            SpringLiquibase liquibase = ((SpringLiquibase) bean);
+
+            PoolConfiguration poolProperties = dataSource.getPoolProperties();
+
+            Map<String, String> params = new HashMap<String, String>() {
+                {
```
[the remark this answers is not in the document]
Comment (uweschaefer) [reply factcast-store-pgsql/src/main/java/org/factcast/store/pgsql/internal/LiquibaseChangelogParamsForwarder.java]: i tend to agree, can we just do a params.put instead?

**Comment by uweschaefer** [comment]
approved. nice one. please check the codacy issue above and update the docs, but then feel free to merge it.
Thanks @OOlsen

**Comment by uweschaefer** [comment]
going to resolve the conflicts for you... just a sec...

**Comment by OOlsen** [comment]
"Codacy/PR Quality Review — Codacy was unable to analyse your pull request."

Is that a kind of configuration problem?
But i resolved the one quality issue, so this pr should be fine regarding this.

**Comment by uweschaefer** [comment]
thx @OOlsen

**Comment by mweirauch** [comment]
I just realized this bombards my ongoing plan to secretly get rid of the tomcat-jdbc pool which is now directly referenced again in the `LiquibaseChangelogParamForwarder`. ;)

**Comment by uweschaefer** [comment]
crap! i'd love to see that direct dep to tomcat-jdbc go.

**Pull request merged**